world.go

```diff
@@ -0,0 +1,258 @@
+	world.log = logger
+	return &world
+}
+
+// GetNodeRecord returns the peer's node record
+// NOTE: do not modify the contents of the returned record! not thread safe
+func (world *World) GetNodeRecord(ID peer.ID) (record *NodeRecord) {
+	world.lk.RLock()
+	defer world.lk.RUnlock()
+	record = world.nodes[ID]
+	return
+}
+
+// SetNodeHolding marks a node as holding a particular hash
+func (world *World) SetNodeHolding(ID peer.ID, hash Hash) (err error) {
+	world.log.Logf("Setting Holding for %v of holding %v nodes:%v\n", ID, hash, world.nodes)
+	world.lk.Lock()
+	defer world.lk.Unlock()
+	record := world.nodes[ID]
+	if record == nil {
+		err = ErrNodeNotFound
+		return
+	}
+	record.IsHolding[hash] = true
+	return
+}
+
+// IsHolding returns whether a node is holding a particular hash
+func (world *World) IsHolding(ID peer.ID, hash Hash) (holding bool, err error) {
+	world.lk.RLock()
+	defer world.lk.RUnlock()
+	world.log.Logf("Looking to see if %v is holding %v\n", ID, hash)
+	world.log.Logf("NODES:%v\n", world.nodes)
+	record := world.nodes[ID]
+	if record == nil {
+		err = ErrNodeNotFound
+		return
+	}
+	holding = record.IsHolding[hash]
+	return
+}
+
+// AllNodes returns a list of all the nodes in the world model.
+func (world *World) AllNodes() (nodes []peer.ID, err error) {
+	world.lk.RLock()
+	defer world.lk.RUnlock()
+	nodes, err = world.allNodes()
+	return
+}
+
+func (world *World) allNodes() (nodes []peer.ID, err error) {
+	nodes = make([]peer.ID, len(world.nodes))
+
+	i := 0
+	for k := range world.nodes {
+		nodes[i] = k
+		i++
+	}
+	return
+}
+
+// AddNode adds a node to the world model
+func (world *World) AddNode(pi pstore.PeerInfo, pubKey ic.PubKey) (err error) {
+	world.lk.Lock()
+	defer world.lk.Unlock()
+	rec := NodeRecord{PeerInfo: pi, PubKey: pubKey, IsHolding: make(map[Hash]bool)}
+	world.nodes[pi.ID] = &rec
+	return
+}
+
+// NodesByHash returns a sorted list of peers, including "me" by distance from a hash
+func (world *World) nodesByHash(hash Hash) (nodes []peer.ID, err error) {
+	nodes, err = world.allNodes()
+	if err != nil {
+		return
+	}
+	nodes = append(nodes, world.me)
+	nodes = SortClosestPeers(nodes, hash)
+	return
+}
+
+/*
+func (world *World) NodeRecordsByHash(hash Hash) (records []*NodeRecord, err error) {
+
+	records = make([]*NodeRecord, len(nodes))
+	i := 0
+	for _, id := range nodes {
+		records[i] = world.nodes[id]
+		i++
+	}
+	return
+}*/
+
+// UpdateResponsible calculates the list of nodes believed to be responsible for a given hash
+// note that if redundancy is 0 the assumption is that all nodes are responsible
+func (world *World) UpdateResponsible(hash Hash, redundancy int) (responsible bool, err error) {
+	world.lk.Lock()
+	defer world.lk.Unlock()
+	var nodes []peer.ID
+	if redundancy == 0 {
+		world.responsible[hash] = nil
+		responsible = true
+	} else if redundancy > 1 {
+		nodes, err = world.nodesByHash(hash)
+		if err != nil {
+			return
+		}
+		// TODO add in resilince calculations with uptime
```

Comment:
can we put issues in github rather than TODO in comments?

or, if we want to keep the TODO, can it look like:

```go
// @TODO blah blah blah
// @see https://github.com/holochain/holochain-proto/issues/1234
```

Reply:
added some tickets!